File: 050_powx_n/Solution2/Solution.py
class Solution:
    def myPow(self, x: float, n: int) -> float:
        def quick_mul(x: float, n: int) -> float:
            result = 1
            while n > 0:
                if n % 2 == 1: result *= x
                x *= x
                n //= 2
            return result

        return quick_mul(x, n) if n >= 0 else 1 / quick_mul(x, -n)


if __name__ == '__main__':
    print(Solution().myPow(2, 10))

File: 050_powx_n/Solution2/Solution.java
class Solution {
    public double quickMul(double x, long n) {
        double result = 1.0;
        while (n >= 1) {
            if (n % 2 == 1) result *= x;
            x *= x;
            n /= 2;
        }

        return result;
    }

    public double myPow(double x, int n) {
        return n >= 0 ? quickMul(x, n) : 1.0 / quickMul(x, -(long) n);
    }

    public static void main(String[] args) {
        System.out.println(new Solution().myPow(2, -2147483648));
    }
}
